service/history/workflow/mutable_state_impl.go

```diff
@@ -1356,6 +1357,41 @@ func (ms *MutableStateImpl) GetCompletionEvent(
 			// can cause task processing side to fail silently
 			return nil, ErrMissingWorkflowCompletionEvent
 		}
+
+		// Certain terminated workflows have an incorrect completionEventBatchId recorded which
+		// prevents the completion event from being loaded outside of cache.
+		//
+		// If we get back an internal error, attempt to search history (most recent
+		// events first) to find the completion event. Event will be earlier than
+		// the recorded CompletionEventBatchID, and should be a part of the same batch.
+		//
+		// see #inc-791-2024-07-16-high-frontend-cpu-on-s-cd046
+		if common.IsInternalError(err) &&
+			ms.executionState.Status == enumspb.WORKFLOW_EXECUTION_STATUS_TERMINATED {
+			_, txID := ms.GetLastFirstEventIDTxnID()
+			resp, err := ms.shard.GetExecutionManager().ReadHistoryBranchReverse(ctx, &persistence.ReadHistoryBranchReverseRequest{
+				ShardID:                ms.shard.GetShardID(),
+				BranchToken:            currentBranchToken,
+				MaxEventID:             nextEventID, // looking for an event in the most recent batch
+				PageSize:               1,
+				LastFirstTransactionID: txID,
+				NextPageToken:          []byte{},
+			})
+			if err != nil {
+				return nil, err
+			}
+
+			for _, event := range resp.HistoryEvents {
+				// this only applies to terminated workflows whose ultimate WFT had been failed
+				if event.EventType == enumspb.EVENT_TYPE_WORKFLOW_TASK_FAILED {
+					return event, nil
```

Comment:
we need to return the workflow completion event, which is the terminated event.

Reply:
fixed, thanks!